internal/metrics/metrics.go

```diff
@@ -0,0 +1,199 @@
+package metrics
+
+import (
+	"context"
+	"fmt"
+	"git.sr.ht/~spc/go-log"
+	"github.com/jakub-dzon/k4e-operator/models"
+	"github.com/prometheus/prometheus/model/labels"
+	"github.com/prometheus/prometheus/storage"
+	"github.com/prometheus/prometheus/tsdb"
+	"os"
+	"path"
+	"sync"
+	"time"
+)
+
+const (
+	// DefaultRetentionDuration specifies 7 days data retention
+	DefaultRetentionDuration = 7 * 24 * time.Hour
+	// DefaultMaxBytes specifies no block data size limit
+	DefaultMaxBytes = int64(0)
+)
+
+type DataPoint struct {
+	Time  int64
+	Value float64
+}
+type Series struct {
+	Labels     map[string]string
+	DataPoints []DataPoint
+}
+
+//go:generate mockgen -package=metrics -destination=storage_mock.go . Storage
+type Storage interface {
+	Deregister() error
+	GetMetricsForTimeRange(tMin time.Time, tMax time.Time) ([]Series, error)
+	AddMetric(value float64, labelsMap map[string]string) error
+}
+
+type TSDB struct {
+	appliedOptions *tsdb.Options
+	db             *tsdb.DB
+	dbLock         sync.RWMutex
+}
+
+func NewTSDB(dataDir string) (*TSDB, error) {
+	metricsDir := path.Join(dataDir, "metrics")
+	if err := os.MkdirAll(metricsDir, 0755); err != nil {
+		return nil, fmt.Errorf("cannot create directory: %w", err)
+	}
+	opts := tsdb.DefaultOptions()
+
+	opts.RetentionDuration = int64(DefaultRetentionDuration / time.Millisecond)
+	opts.MaxBytes = DefaultMaxBytes
+
+	db, err := tsdb.Open(metricsDir, nil, nil, opts, nil)
+	if err != nil {
+		return nil, err
+	}
+	return &TSDB{
+		db:             db,
+		dbLock:         sync.RWMutex{},
+		appliedOptions: opts,
+	}, err
+}
+
+func (t *TSDB) Deregister() error {
+	metricsDir := t.db.Dir()
+	t.dbLock.Lock()
+	defer t.dbLock.Unlock()
+
+	err := t.db.Close()
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+
+	err = os.RemoveAll(metricsDir)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	return nil
+}
+
+func (t *TSDB) GetMetricsForTimeRange(tMin time.Time, tMax time.Time) ([]Series, error) {
+	log.Tracef("Getting metrics for %v - %v", tMin, tMax)
+	q, err := t.getDB().Querier(nil, toDbTime(tMin), toDbTime(tMax))
+	if err != nil {
+		return nil, err
+	}
+	defer func() {
+		err = q.Close()
+		if err != nil {
+			log.Error(err)
+		}
+	}()
+
+	matchers, err := getAllLabelsMatchers(q)
+	if err != nil {
+		return nil, err
+	}
+	seriesSet := q.Select(false, nil, matchers...)
+	if seriesSet.Err() != nil {
+		log.Error(seriesSet.Err())
+		return nil, seriesSet.Err()
+	}
+
+	return toSeries(seriesSet), nil
+}
+
+func toSeries(seriesSet storage.SeriesSet) []Series {
+	var allSeries []Series
+	for {
+		if !seriesSet.Next() {
+			break
+		}
+		s := seriesSet.At()
```

Comment:
the seriesSet is sorted. our map is not. that's ok?

Reply:
IMO for labels it's OK not to be sorted as order of labels doesn't have any significance.